File: tutorials/tour-of-restate-java/src/main/java/dev/restate/tour/part4/CheckoutService.java
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Tour of Restate Java,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

package dev.restate.tour.part4;

import dev.restate.sdk.Context;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import dev.restate.sdk.common.CoreSerdes;
import dev.restate.tour.auxiliary.CheckoutRequest;
import dev.restate.tour.auxiliary.EmailClient;
import dev.restate.tour.auxiliary.PaymentClient;

@Service
public class CheckoutService {
    // <start_checkout>
    @Handler
    public boolean handle(Context ctx, CheckoutRequest request) {
        // <start_side_effects>
        double totalPrice = request.getTickets().size() * 40.0;

        String idempotencyKey = ctx.random().nextUUID().toString();
        boolean success = ctx.run(CoreSerdes.JSON_BOOLEAN, () -> PaymentClient.get().call(idempotencyKey, totalPrice));
        // <end_side_effects>

        if (success) {
            ctx.run(()-> EmailClient.get().notifyUserOfPaymentSuccess(request.getUserId()));
        } else {
            ctx.run(() -> EmailClient.get().notifyUserOfPaymentFailure(request.getUserId()));
        }

        return success;
    }
    // <end_checkout>
}

File: python/tutorials/tour-of-restate-python/tour/part1/checkout_service.py
# Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
#
# This file is part of the Restate examples,
# which is released under the MIT license.
#
# You can find a copy of the license in the file LICENSE
# in the root directory of this repository or package or at
# https://github.com/restatedev/examples/

from typing import TypedDict, List
from restate.context import ObjectContext, Serde
from restate.service import Service

from tour.auxiliary.email_client import EmailClient
from tour.auxiliary.payment_client import PaymentClient


class Order(TypedDict):
    user_id: str
    tickets: List[str]


payment_client = PaymentClient()
email_client = EmailClient()

checkout = Service("CheckoutService")


# <start_checkout>
@checkout.handler()
async def handle(ctx: ObjectContext, order: Order) -> bool:
    return True
# <end_checkout>
